cmd/dra-example-kubeletplugin/state.go

```diff
@@ -173,3 +206,97 @@ func (s *DeviceState) prepareDevices(claim *resourceapi.ResourceClaim) (Prepared
+func (pds PreparedDevices) GetDevices() []*drapbv1.Device {
+	var devices []*drapbv1.Device
+	for _, pd := range pds {
+		devices = append(devices, &pd.Device)
+	}
+	return devices
+}
+
+// ptrImplementsRuntimeObject is a helper interface for GetOpaqueDeviceConfig.
+//
+// It enforces that a pointer to the type parameter passed to
+// GetOpaqueDeviceConfig implements the runtime.Object interface as well as a
+// Validate() function.
+type ptrImplementsRuntimeObject[T any] interface {
+	*T
+	runtime.Object
+	Validate() error
+}
+
+// GetOpaqueDeviceConfig returns the config with the highest precedence specified for a device request in a resource claim.
+//
+// Configs can either be attached to the resource claim itself or to the
+// device class associated with the request. Configs found directly in the
+// resource claim take precedence over configs found in the device class.
+// Moreover, configs found earlier in the list of configs attached to either a
+// resource claim or a device class take precedence over configs found later in
+// their corresponding list.
+//
+// Only the config with the highest precedence in the order specified above will
+// be returned. All other configs will be ignored. If no config is found, the
+// default passed to this function will be returned instead.
+func GetOpaqueDeviceConfig[T any, TI ptrImplementsRuntimeObject[T]](
+	defaultConfig T,
+	decoder runtime.Decoder,
+	driverName string,
+	request string,
+	resourceClaim *resourceapi.ResourceClaim,
+	deviceClasses map[string]*resourceapi.DeviceClass,
+) (T, error) {
+	// Collect all configs in order of reverse precedence.
+	var allConfigs []resourceapi.DeviceConfiguration
+	for _, config := range deviceClasses[request].Spec.Config {
+		allConfigs = append(allConfigs, config.DeviceConfiguration)
+	}
+	for _, config := range resourceClaim.Spec.Devices.Config {
+		if !slices.Contains(config.Requests, request) {
+			continue
+		}
+		allConfigs = append(allConfigs, config.DeviceConfiguration)
+	}
+
+	// Walk all configs from the allConfigs slice in the order they were added.
+	// The latest config matching the GroupVersionKind of the defaultConfig
+	// will be returned. If no configs match, then the original defaultConfig
+	// is returned.
+	candidateConfig := defaultConfig
+	for _, config := range allConfigs {
+		if config.Opaque == nil {
+			continue
+		}
+		if config.Opaque.Driver != driverName {
+			continue
+		}
+
+		c, err := runtime.Decode(decoder, config.Opaque.Parameters.Raw)
+		if err != nil {
+			return *new(T), fmt.Errorf("error decoding config parameters: %w", err)
+		}
+
+		if c.GetObjectKind().GroupVersionKind() != TI(&defaultConfig).GetObjectKind().GroupVersionKind() {
+			continue
+		}
+
+		config, ok := c.(TI)
+		if !ok {
+			return *new(T), fmt.Errorf("error asserting actual runtime object type: %T", c)
+		}
+
+		candidateConfig = *config
+	}
+
+	// Validate the config before returning it.
```

Comment:
Can't the caller do that? Then ptrImplementsRuntimeObject becomes simpler. The caller can also validate that there are no conflicting configs when multiple devices were allocated.

Reply:
Updated.